template.js

```diff
@@ -3,117 +3,102 @@
 // SPDX-License-Identifier: LGPL-3.0-only
 
 const { QL0 } = require('ssb-subset-ql')
+const pickShard = require('ssb-meta-feeds/pick-shard')
+
+function isEmptyObject(obj) {
+  return Object.keys(obj).length === 0
+}
 
 /**
  * Algorithms for the "replication template" object from the SSB config.
  */
 module.exports = class Template {
-  constructor(node) {
-    this._rootNode = node
+  constructor(leafShapes) {
+    /** @type Array{*} */
+    this._leafShapes = leafShapes
   }
 
   hasIndexLeaf() {
-    const indexes = this._find(
-      this._rootNode,
-      (node) => node.feedpurpose === 'indexes'
-    )
-    if (!indexes) return false
-    const leaf = this._find(indexes, (node) => node['$format'] === 'indexed')
-    return !!leaf
+    return this._leafShapes.some((shape) => shape && shape.purpose === 'index')
   }
 
-  _find(node, fn) {
-    if (!node) return null
-    if (fn(node)) return node
+  matchBranch(branch, mainFeedId) {
+    return this._matchBranch(branch, mainFeedId)
+  }
 
-    if (node.subfeeds && Array.isArray(node.subfeeds)) {
-      for (const childNode of node.subfeeds) {
-        if (this._find(childNode, fn)) {
-          return childNode
-        }
-      }
+  _matchBranch(branch, mainFeedId) {
+    if (!branch || !Array.isArray(branch)) return false
+    const [rootMF, v1MF, shardMF, leafFeed] = branch
+    switch (branch.length) {
+      case 0:
+        return false
+      case 1:
+        return true
+      case 2:
+        return true
+      case 3:
+        return this._matchShard(rootMF, shardMF)
+      case 4:
+        return this._matchLeaf(leafFeed, rootMF.id, mainFeedId)
+      default:
+        return false
     }
   }
 
-  matchBranch(branch, mainFeedId) {
-    return this._matchBranch(branch, this._rootNode, mainFeedId)
+  _matchShard(rootMF, shardMF) {
+    if (this._leafShapes.some(isEmptyObject)) return true
+    const wantedShards = this._leafShapes.map((leaf) =>
+      pickShard(rootMF.id, leaf.purpose)
+    )
+    return wantedShards.includes(shardMF.purpose)
```

Comment:
might be a bit easier to read if this was written like in matchLeaf, so: 

```
this._leafShapes.some((leaf) => {
    if (isEmptyObject(leaf)) return true

    const shardPurpose = pickShard(rootMF.id, leaf.purpose)
    return shardPurpose === shardMF.purpose
})
````

What do you think?